ambry-replication/src/main/java/com/github/ambry/replication/BackupCheckerThread.java

```diff
@@ -0,0 +1,165 @@
+ * If the local store was recovered from a backup, then we are effectively checking the blob state in backup.
+ * The backup is a union of all on-prem replicas. So the latest state of the blob should be reflected in the backup.
+ * There are two possibilities:
+ * 1> The backup may be lagging behind the replica and if we recover from such a lagging backup, then we'd be recovering
+ * to a point in time in the past.
+ * 2> The backup may be ahead of the replica. This can happen if the replica is lagging behind its peers. If we recover
+ * from such a backup, we'd still be consistent from the user's point of view. The lagging replica must catch up with
+ * its peers and this checker will detect such lagging replicas as well.
+ * TODO: Since we are just checking for existence of a blob in local-store or backup, we don't need to fetch the entire
+ * TODO: blob from on-prem server. Consider overriding that method and getting just a checksum from on-prem server.
+ */
+public class BackupCheckerThread extends ReplicaThread {
+
+  private final Logger logger = LoggerFactory.getLogger(BackupCheckerThread.class);
+
+  public BackupCheckerThread(String threadName, FindTokenHelper findTokenHelper, ClusterMap clusterMap,
+      AtomicInteger correlationIdGenerator, DataNodeId dataNodeId, ConnectionPool connectionPool,
+      ReplicationConfig replicationConfig, ReplicationMetrics replicationMetrics, NotificationSystem notification,
+      StoreKeyConverter storeKeyConverter, Transformer transformer, MetricRegistry metricRegistry,
+      boolean replicatingOverSsl, String datacenterName, ResponseHandler responseHandler, Time time,
+      ReplicaSyncUpManager replicaSyncUpManager, Predicate<MessageInfo> skipPredicate,
+      ReplicationManager.LeaderBasedReplicationAdmin leaderBasedReplicationAdmin) {
+    super(threadName, findTokenHelper, clusterMap, correlationIdGenerator, dataNodeId, connectionPool,
+        replicationConfig, replicationMetrics, notification, storeKeyConverter, transformer, metricRegistry,
+        replicatingOverSsl, datacenterName, responseHandler, time, replicaSyncUpManager, skipPredicate,
+        leaderBasedReplicationAdmin);
+  }
+
+  /**
+   *  This checks for GDPR compliance.
+   *  If a blob has been deleted from on-prem servers, it must be absent in backups.
+   *  If it is in TEMP or PERM state, then backup is behind servers & DELETE has not propagated to the backup.
+   * @param onPremBlob the {@link MessageInfo} that will be transformed into a delete-operation
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyDelete(MessageInfo onPremBlob, RemoteReplicaInfo remoteReplicaInfo) {
+    EnumSet<MessageInfoType> acceptableBackupBlobStates = EnumSet.noneOf(MessageInfoType.class);
+    EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.of(StoreErrorCodes.ID_Deleted,
```

Comment:
yes, applyXXX in this context is actually verifyXXX. I cannot rename it because i am extending the replicaThread and overriding its methods. But I need to read through the several cases of replicaThread to see under what conditions does it come here.

Reply:
Yes, you are right. I just read the replicaThread code and all the cases and your comment makes more sense now. If we are here, then that means the blob is present in the local store but not in deleted state. This makes my coding much simpler.